lib/route/src/Obelisk/Route.hs

```diff
@@ -526,6 +546,134 @@ checkEnum1EncoderFunc f = do
+  Nothing -> Nothing
+  Just auth -> Just $ Domain $ T.pack $ "//" <> URI.uriRegName auth <> URI.uriPort auth
+
+domainToString :: Domain -> String
+domainToString = T.unpack . unDomain
+
+type DomainPageName = (Domain, PageName)
+
+data DomainConfig a = DomainConfig
+  { _domainConfig_toConcrete :: a -> ConcreteDomain
+  , _domainConfig_uris :: [URI]
+  }
+
+-- | Return all URIs in the store
+domainConfigURIs :: DomainConfig a -> [URI]
+domainConfigURIs = _domainConfig_uris
+
+-- | Decode a domain config
+decodeDomainConfig :: forall domain. (Universe domain, Ord domain, Show domain) => ByteString -> Either Text (DomainConfig domain)
+decodeDomainConfig rawConfig = first (T.unlines . (:exampleFormat)) $ do
+  let
+    stripped = T.strip (T.decodeUtf8 rawConfig)
+    pairs = T.breakOn " " . T.strip <$> T.lines stripped
+    parseURI t = case URI.parseURI $ T.unpack $ T.strip t of
+      Just uri -> Right uri
+      Nothing -> Left $ "decodeDomainConfig: Couldn't parse entry in common/route as URI; value read was: " <> tshow (T.strip t)
+    allDomains = Map.fromList $ flip fmap universe $ \(domain :: domain) -> (T.pack $ show domain, domain)
+  domainToURI <- fmap Map.fromList $ for pairs $ \(tagString, uriString) -> do
+    tag' <- case Map.lookup (T.strip tagString) allDomains of
+      Nothing -> Left $ "decodeDomainConfig: common/route config has unknown tag: " <> tagString
+      Just t -> Right t
+    uri <- parseURI uriString
+    domain <- case uriToDomain uri of
+      Nothing -> Left $ "decodeDomainConfig: URI " <> tshow uri <> " could not be converted to a domain."
+      Just domain -> Right domain
+    Right (tag', (uri, domain))
+  when (Map.size domainToURI < Map.size allDomains) $
+    Left $ T.unwords
+      [ "decodeDomainConfig: Not enough items in common/route config. Missing entries for:"
+      , tshow $ Set.toList $ Set.fromList universe Set.\\ Map.keysSet domainToURI
+      ]
+  when (Set.size (Set.fromList $ fmap snd $ Map.elems domainToURI) < Map.size allDomains) $
+    Left "decodeDomainConfig: Domains given in common/route config overlap. Each domain must be unique."
+  pure $ DomainConfig
+    { _domainConfig_toConcrete = \domain -> case Map.lookup domain domainToURI of
+      Nothing -> error $ "decodeDomainConfig: Domain " <> show domain <> " was missing from the map. This should be impossible!"
+      Just (_, Domain d) -> ConcreteDomain d
+    , _domainConfig_uris = fmap fst $ Map.elems domainToURI
+    }
+  where
+    exampleFormat =
+      [ ""
+      , "  Example common/route config for an app with a single domain:"
+      , ""
+      , "() https://mydomain-a.com"
+      , ""
+      , "  Or for apps with multiple custom domains:"
+      , ""
+      , "MyDomain_A https://mydomain-a.com"
+      , "MyDomain_B https://mydomain-b.com"
+      , ""
+      , "  Or for local dev work, you can specify different ports:"
+      , ""
+      , "MyDomain_A http://localhost:8000"
+      , "MyDomain_B http://localhost:8001"
```

Comment:
Does this support a variable number of domains (e.g. database driven domains)?

Reply:
No, that was specifically excluded for simplicity. It’s worth think about how we would extend this to support that though.

Reply:
I probably haven't ingested what this is doing well enough to properly comment, but it seems like the domain doesn't need to be a separate type variable in the route type and instead could be reduced to nothing more than an encoder where the URL type is enriched to include this information.

Reply:
In particular that seems like it would support variable number of domains by default.

Reply:
We do need to check that there is no overlap, which is currently done by universe.

Reply:
I see. What's the benefit of adding the type parameter? I would imagine that an encoder could choose to map domains to any arbitrary part of the encoding.

Reply:
I think it would be good if we can reduce the assumptions that Obelisk makes here.  For instance, what if someone wanted to just "decode" the domain and port using `id`, so that their app receives (Text, Word16) and can do whatever it likes with it?

Reply:
@3noch it's currently used so that the Universe instance for ObeliskRoute_Resource can generate routes for each domain, which is later used in overlap checks.

Reply:
@ryantrinkle I think that's still possible, but you'd need to use the DomainConfig to get that information. I'm not sure of a way to get rid of that...

Reply:
@tomsmalley Can't ObeliskRoute_Resource just apply to all domains?  It doesn't seem like it should need to enumerate the domains, just have an appropriate Encoder.

Reply:
@ryantrinkle Ah, I thought you were against that based on one of our conversations, but I must've misunderstood.